### PR TITLE
Calib matrix index

### PR DESCRIPTION
**Pull request by asedlmayr**
Downliink calibration index updates
PacketTxRx changes for the client

### CODE REVIEW

**Review by di-mu — APPROVED** [review]

**Review by di-mu — APPROVED** [review]

File: src/agora/dozf.cc [hunk 1]
```diff
@@ -391,39 +412,35 @@ void DoZF::ZfFreqOrthogonal(size_t tag) {
...
-        reinterpret_cast<arma::cx_float*>(calib_gather_buffer_),
-        cfg_->BfAntNum(), false);
     size_t frame_cal_slot = kFrameWnd - 1;
-    size_t frame_cal_slot_prev = kFrameWnd - 1;
-    if (cfg_->Frame().IsRecCalEnabled() && (frame_id >= TX_FRAME_DELTA)) {
-      size_t frame_grp_id = (frame_id - TX_FRAME_DELTA) / cfg_->AntGroupNum();
+    size_t frame_cal_slot_prev = kFrameWnd - 2;
+    if (cfg_->Frame().IsRecCalEnabled()) {
+      const size_t frames_to_complete = cfg_->RecipCalFrameCnt();
+      const size_t current_cal_index = frame_id / frames_to_complete;
```
Comment (rdoosty) [inline src/agora/dozf.cc:419]: Should we make these two lines a function too since it is repeated throughout?

File: src/common/config.h [hunk 2]
```diff
@@ -117,30 +115,52 @@ class Config {
...
     return (tx_ant);
   };
 
   // Returns the cal index if ant tx dl cal pilots this frame
   // SIZE_MAX otherwise
-  size_t RecipCalUlRxIndex(size_t frame_id, size_t ant) const {
+  inline size_t RecipCalUlRxIndex(size_t frame_id, size_t ant) const {
+    const size_t num_frames_for_full_cal = RecipCalFrameCnt();
+    const size_t num_cal_per_idx = frame_.NumDLCalSyms();
+    const size_t cal_offset = (frame_id % num_frames_for_full_cal);
```
Comment (rdoosty) [inline src/common/config.h:131]: Use the function here too.

File: src/agora/dozf.cc [hunk 3]
```diff
@@ -152,75 +161,85 @@ float DoZF::ComputePrecoder(const arma::cx_fmat& mat_csi,
...
-    if (frame_id >= TX_FRAME_DELTA + cfg_->AntGroupNum()) {
-      frame_cal_slot_prev = (frame_grp_id + kFrameWnd - 2) % kFrameWnd;
+// Called for each frame_id / sc_id
+// Updates calib_sc_vec
+void DoZF::ComputeCalib(size_t frame_id, size_t sc_id,
+                        arma::cx_fvec& calib_sc_vec) {
+  const size_t frames_to_complete = cfg_->RecipCalFrameCnt();
+  if (cfg_->Frame().IsRecCalEnabled() && (frame_id >= frames_to_complete)) {
+    const size_t current_cal_index = frame_id / frames_to_complete;
+    const bool frame_update = ((frame_id % frames_to_complete) == 0);
```
[the remark this answers is not in the document]
Comment (asedlmayr) [reply src/agora/dozf.cc:171]: I agree with this point.  I reworked some things around this but left the update condition the same as I didn't want to run a full regression suite.

File: src/agora/dozf.cc (hunk 1 above)
Comment (asedlmayr) [reply src/agora/dozf.cc]: Reworked

File: src/common/config.h (hunk 2 above)
Comment (asedlmayr) [reply src/common/config.h:141]: Reworked

**Review by rdoosty — APPROVED** [review]

**Pull request merged**